File: src/back-end/FunChess/src/FunChess.Core.Chess/Constants/BoardConstants.cs
using FunChess.Core.Chess.Enums;

namespace FunChess.Core.Chess.Constants;

public static class BoardConstants
{
    public const int Length = 8;
    public const int TotalSize = Length * Length;
    public const int MaxIndex = Length - 1;
    public const int MinIndex = 0;

    public static Cell[] GenerateDefaultBoard()
    {
        var board = new Cell[TotalSize];
        Array.Copy(InitialBoard, board, TotalSize);

        return board;
    }

    public static ReadOnlyMemory<Cell> GetInitialBoard() => InitialBoard;

    private static readonly Cell[] InitialBoard =
    {
        Cell.Get(Piece.Rook, Team.White), Cell.Get(Piece.Knight, Team.White), Cell.Get(Piece.Bishop, Team.White), Cell.Get(Piece.Queen, Team.White),
        Cell.Get(Piece.King, Team.White), Cell.Get(Piece.Bishop, Team.White), Cell.Get(Piece.Knight, Team.White), Cell.Get(Piece.Rook, Team.White),
        Cell.Get(Piece.Pawn, Team.White), Cell.Get(Piece.Pawn, Team.White), Cell.Get(Piece.Pawn, Team.White), Cell.Get(Piece.Pawn, Team.White),
        Cell.Get(Piece.Pawn, Team.White), Cell.Get(Piece.Pawn, Team.White), Cell.Get(Piece.Pawn, Team.White), Cell.Get(Piece.Pawn, Team.White),
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Get(Piece.Pawn, Team.Black), Cell.Get(Piece.Pawn, Team.Black), Cell.Get(Piece.Pawn, Team.Black), Cell.Get(Piece.Pawn, Team.Black),
        Cell.Get(Piece.Pawn, Team.Black), Cell.Get(Piece.Pawn, Team.Black), Cell.Get(Piece.Pawn, Team.Black), Cell.Get(Piece.Pawn, Team.Black),
        Cell.Get(Piece.Rook, Team.Black), Cell.Get(Piece.Knight, Team.Black), Cell.Get(Piece.Bishop, Team.Black), Cell.Get(Piece.Queen, Team.Black),
        Cell.Get(Piece.King, Team.Black), Cell.Get(Piece.Bishop, Team.Black), Cell.Get(Piece.Knight, Team.Black), Cell.Get(Piece.Rook, Team.Black),
    };
}

File: src/front-end/fun-chess-app/src/core/chess/constants/board-constants.ts
import Cell from "../Cell";
import Team from "../enums/Team";
import Bishop from "../pieces/Bishop";
import King from "../pieces/King";
import Knight from "../pieces/Knight";
import Pawn from "../pieces/Pawn";
import Queen from "../pieces/Queen";
import Rook from "../pieces/Rook";

export const BOARD_LENGTH = 8;

export const BOARD_TOTAL_SIZE = BOARD_LENGTH * BOARD_LENGTH;

export const MAX_INDEX = BOARD_LENGTH - 1;

export const MIN_INDEX = 0;

export function getInitialBoard(): Cell[] {
    return [
        Cell.get(Rook.instance, Team.White), Cell.get(Knight.instance, Team.White), Cell.get(Bishop.instance, Team.White), Cell.get(Queen.instance, Team.White),
        Cell.get(King.instance, Team.White), Cell.get(Bishop.instance, Team.White), Cell.get(Knight.instance, Team.White), Cell.get(Rook.instance, Team.White),
        Cell.get(Pawn.instance, Team.White), Cell.get(Pawn.instance, Team.White), Cell.get(Pawn.instance, Team.White), Cell.get(Pawn.instance, Team.White),
        Cell.get(Pawn.instance, Team.White), Cell.get(Pawn.instance, Team.White), Cell.get(Pawn.instance, Team.White), Cell.get(Pawn.instance, Team.White),
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.Empty, Cell.Empty, Cell.Empty, Cell.Empty,
        Cell.get(Pawn.instance, Team.Black), Cell.get(Pawn.instance, Team.Black), Cell.get(Pawn.instance, Team.Black), Cell.get(Pawn.instance, Team.Black),
        Cell.get(Pawn.instance, Team.Black), Cell.get(Pawn.instance, Team.Black), Cell.get(Pawn.instance, Team.Black), Cell.get(Pawn.instance, Team.Black),
        Cell.get(Rook.instance, Team.Black), Cell.get(Knight.instance, Team.Black), Cell.get(Bishop.instance, Team.Black), Cell.get(Queen.instance, Team.Black),
        Cell.get(King.instance, Team.Black), Cell.get(Bishop.instance, Team.Black), Cell.get(Knight.instance, Team.Black), Cell.get(Rook.instance, Team.Black)
    ];
}